README.md
# UTXO Prometheus Exporter

A [Prometheus] exporter for UTXO blockchain nodes written in python and packaged for running as a container.

A rudimentary Grafana [dashboard] is available in the [`dashboard/bitcoin-grafana.json`](dashboard/bitcoin-grafana.json)
file.

The main script is a modified version of [`bitcoin-monitor.py`][source-gist], updated to remove the need for the
`bitcoin-cli` binary, packaged into a [Docker image][docker-image], and expanded to export additional metrics.

[Prometheus]: https://github.com/prometheus/prometheus
[docker-image]: https://hub.docker.com/r/jvstein/bitcoin-prometheus-exporter

[source-gist]: https://gist.github.com/ageis/a0623ae6ec9cfc72e5cb6bde5754ab1f
[python-bitcoinlib]: https://github.com/petertodd/python-bitcoinlib
[dashboard]: https://grafana.com/grafana/dashboards/11274

## Usage

### Run without docker

1. You need to have python >= 3.8 and [pipenv]
2. Install dependencies:

    pipenv install

3. Load the environment:

    pipenv shell

4. Run the script:

    UTXO_NODE_RPC_HOST=bitcoin-node-address \
    UTXO_NODE_RPC_USER=rpc-username \
    UTXO_NODE_RPC_PASSWORD=rpc-password \
    python3 utxo_prometheus_exporter.py

[pipenv]: https://pipenv.pypa.io/en/latest/

### Run with docker
```sh
docker run \
    --name=utxo-prometheus-exporter \
    -p 9332:9332 \
    -e UTXO_NODE_BLOCKCHAIN_NAME=bitcoin
    -e UTXO_NODE_RPC_HOST=bitcoin-node \
    -e UTXO_NODE_RPC_USER=alice \
    -e UTXO_NODE_RPC_PASSWORD=DONT_USE_THIS_YOU_WILL_GET_ROBBED_8ak1gI25KFTvjovL3gAM967mies3E= \
    santiment/utxo-prometheus-exporter:latest
```

### Run in Kubernetes

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: litecoin-prometheus-exporter
  labels:
    app: litecoin-prometheus-exporter
spec:
  selector:
    matchLabels:
      app: litecoin-prometheus-exporter
  replicas: 1
  template:
    metadata:
      labels:
        app: litecoin-prometheus-exporter
      annotations:
        prometheus.io/scrape: 'true'
        prometheus.io/port:   '9335'
    spec:
      containers:
        - name: litecoin-prometheus-exporter
          image: santiment/utxo-prometheus-exporter:latest
          imagePullPolicy: Always
          env:
            - name: UTXO_NODE_BLOCKCHAIN_NAME
              value: litecoin
            - name: UTXO_NODE_RPC_HOST
              value: "litecoin-node.default"
            - name: UTXO_NODE_RPC_PORT
              value: "9332"
            - name: UTXO_NODE_RPC_USER
              value: "rpc-user"
            - name: UTXO_NODE_RPC_PASSWORD
              value: "rpc-password"
            - name: METRICS_PORT
              value: "9335"
```

## Basic Testing
There's a [`docker-compose.yml`](docker-compose.yml) file in the repository that references a test bitcoin node. To test changes to the exporter in docker, run the following commands.

```
docker-compose down
docker-compose up --build
```

If you see a lot of `ConnectionRefusedError` errors, run `chmod og+r test-bitcoin.conf`.

# [Change Log](CHANGELOG.md)
See the [`CHANGELOG.md`](CHANGELOG.md) file for changes.


CHANGELOG.md
# Changelog
Changes to the project.

## [1.0.0] - 2022-11-04

**BREAKING CHANGE: All prometheus metrics and env variables has been renamed.**

### Added
- Prometheus metric names have `utxo_node` prefix instead of the old `bitcoin`
- Env variables have `UTXO_NODE` prefix instead of the old `BITCOIN`
- All prometheus metrics have label `blockchain` which can be set with `UTXO_NODE_BLOCKCHAIN_NAME` env variable.
- Configuration has been extracted to separate file - `config.py`
- Prometheus metrics have been extracted to separate file - `prometheus_metrics.py`

## [0.8.0] - 2022-11-02

### Added
- It now support other UTXO blockchains besides bitcoin core
- Ability to control which RPC calls will be made by supplying the following env vars:
* `FETCH_UPTIME`
* `FETCH_MEMINFO`
* `FETCH_BLOCKCHAININFO`
* `FETCH_NETWORKINFO`
* `FETCH_CHAINTIPS`
* `FETCH_MEMPOOLINFO`
* `FETCH_NETTOTALS`
* `FETCH_RPCINFO`
* `FETCH_TXSTATS`
* `FETCH_BANNED`
* `FETCH_SMART_FEES`
* `FETCH_HASHP_BLOCKS`

- RPC calling and error handling was reworked to be per every call so now if a call fails it won't
affect the other ones.


## [0.7.0] - 2022-01-14

**BREAKING CHANGE: Default port number has changed to 9332! Use `METRICS_PORT=8334` to use old port.**

### Added
- Add `bitcoin_rpc_active` stat for number of active RPC calls.
  - [Source](https://github.com/EchterAgo/bitcoin-prometheus-exporter/commit/cc2a804e214556414c7184830166242c34d42457)
- Add `bitcoin_txcount` stat for total number of transactions since the genesis block.
  - [Source 1](https://github.com/EchterAgo/bitcoin-prometheus-exporter/commit/b368138574641e4e26a8a2dfc8be6eede82f4a73)
  - [Source 2](https://github.com/EchterAgo/bitcoin-prometheus-exporter/commit/a9a6b250f463906c1b1f9446d9a689c78c4add6d)
- Support for Bitcoin Cash nodes ([#22][pr-22]).

[pr-22]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/22

### Changed
- Default port changed from `8334` to `9332` to avoid conflicts with Bitcoin Tor port.
- Update metrics on HTTP request, instead of timer ([#12][issue-12]).
  - [Source 1](https://github.com/EchterAgo/bitcoin-prometheus-exporter/commit/c8382240b7a931503dfdd4c8cf89a8415326caf6)
  - [Source 2](https://github.com/EchterAgo/bitcoin-prometheus-exporter/commit/89212072386307fcb6a9f062ee7f958a266b1075)
- Improve performance of block statistics.
  - [Source](https://github.com/EchterAgo/bitcoin-prometheus-exporter/commit/9c018bf081bfdc604af03d8dedd125197401b2de)
- The number of blocks for the `bitcoin_hashps` metrics is now configurable via `HASHPS_BLOCKS` ([#22][pr-22]).

[issue-12]: https://github.com/jvstein/bitcoin-prometheus-exporter/issues/12


## [0.6.0] - 2021-06-05

### Added
- Support changing bind address with `METRICS_ADDR` environment variable ([#11][pr-11]).
- Add `requirements.txt` file.
- Set default `bad_reason` to "manually added" to support Bitcoin Core 0.20.1 ([#16][pr-16]).
- Remove premature URL encoding of bitcoind rpc credentials ([#18][pr-18]).
- New metrics for Bitcoin Core 0.21.0.
- Support for ARM docker image builds via `buildx`.

[pr-11]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/11
[pr-16]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/16
[pr-18]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/18

### Fixed
- Fix port number in README ([#17][pr-17]).

[pr-17]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/17

## Changed
- Modify type annotations to run on python 3.5.


## [0.5.0] - 2020-02-10

### Fixed
- Avoid crash on `socket.timeout` errors. Retry the request in that case.

### Changed
- Switch to python 3.8 and alpine for base image.
- Update docker container to use `nobody` instead of default `root` account.
- Update shebang to use PATH ([#6][pr-6]).
- Support loading bitcoin config from `BITCOIN_CONF_PATH` environment variable ([#7][pr-7]).
- Reuse the `Proxy` RPC client to avoid repeatedly looking for config file.
- Config file examples in the docker-compose file.
- Retry on `ConnectionError`, not just `ConnectionRefusedError` ([#9][pr-9]).
- Pass `TIMEOUT` environment value to bitcoin client as well as retry library.
- Rely on python-bitcoinlib to handle bitcoin config location detection ([#10][pr-10]).

[pr-6]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/6
[pr-7]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/7
[pr-9]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/9
[pr-10]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/10


## [0.4.0] - 2020-01-05

### Added
- New counter metric `bitcoin_exporter_process_time_total` for time spent refreshing the metrics.
- New `bitcoin_verification_progress` metric to track chain verification progress ([#5][pr-5]).
- Use `logging` for output messages and improve level of output for errors ([#4][issue-4]).
- Add docker-compose config with basic setup for testing regressions.

[pr-5]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/5
[issue-4]: https://github.com/jvstein/bitcoin-prometheus-exporter/issues/4

### Changed
- Retry failed RPC calls with exponential timeout using riprova and keep track of retry exceptions using new
  `bitcoin_exporter_errors` metric.
- Improved error message when credentials are incorrect.
- Make smartfee metrics configurable using `SMARTFEE_BLOCKS` environment variable.
- Update script shebang to use PATH ([#6][pr-6])
- Prefer the `$HOME/.bitcoin/bitcoin.conf` file over `BITCOIN_RPC_HOST`, `BITCOIN_RPC_USER`, ... if it exists ([#7][pr-7]).
- Add pre-commit hooks for catching style and code issues.
- Added type annotations and no longer attempting to support less than Python 3.7.

[pr-6]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/6
[pr-7]: https://github.com/jvstein/bitcoin-prometheus-exporter/pull/7

### Fixed
- Avoid crashing on node restart by ignoring `bitcoin.rpc.InWarmupError` exception.
- Prevent KeyError when smartfee values are not calculable ([#2][issue-2]).

prometheus_metrics.py
from decimal import Decimal
from prometheus_client import Gauge, Counter

METRIC_PREFIX = "utxo_node"

UTXO_NODE_BLOCKS = Gauge(f"{METRIC_PREFIX}_blocks", "Block height", labelnames=["blockchain"])
UTXO_NODE_DIFFICULTY = Gauge(f"{METRIC_PREFIX}_difficulty", "Difficulty", labelnames=["blockchain"])
UTXO_NODE_PEERS = Gauge(f"{METRIC_PREFIX}_peers", "Number of peers", labelnames=["blockchain"])
UTXO_NODE_CONN_IN = Gauge(f"{METRIC_PREFIX}_conn_in", "Number of connections in", labelnames=["blockchain"])
UTXO_NODE_CONN_OUT = Gauge(f"{METRIC_PREFIX}_conn_out", "Number of connections out", labelnames=["blockchain"])

UTXO_NODE_HASHPS_GAUGES = {}  # type: Dict[int, Gauge]
UTXO_NODE_ESTIMATED_SMART_FEE_GAUGES = {}  # type: Dict[int, Gauge]

UTXO_NODE_WARNINGS = Counter(
    f"{METRIC_PREFIX}_warnings",
    "Number of network or blockchain warnings detected",
    labelnames=["blockchain"]
)
UTXO_NODE_UPTIME = Gauge(
    f"{METRIC_PREFIX}_uptime",
    "Number of seconds the node has been running",
    labelnames=["blockchain"]
)

UTXO_NODE_MEMINFO_USED = Gauge(f"{METRIC_PREFIX}_meminfo_used", "Number of bytes used", labelnames=["blockchain"])
UTXO_NODE_MEMINFO_FREE = Gauge(f"{METRIC_PREFIX}_meminfo_free", "Number of bytes available", labelnames=["blockchain"])
UTXO_NODE_MEMINFO_TOTAL = Gauge(f"{METRIC_PREFIX}_meminfo_total", "Number of bytes managed", labelnames=["blockchain"])
UTXO_NODE_MEMINFO_LOCKED = Gauge(f"{METRIC_PREFIX}_meminfo_locked", "Number of bytes locked", labelnames=["blockchain"])
UTXO_NODE_MEMINFO_CHUNKS_USED = Gauge(f"{METRIC_PREFIX}_meminfo_chunks_used", "Number of allocated chunks", labelnames=["blockchain"])
UTXO_NODE_MEMINFO_CHUNKS_FREE = Gauge(f"{METRIC_PREFIX}_meminfo_chunks_free", "Number of unused chunks", labelnames=["blockchain"])

UTXO_NODE_MEMPOOL_BYTES = Gauge(f"{METRIC_PREFIX}_mempool_bytes", "Size of mempool in bytes", labelnames=["blockchain"])
UTXO_NODE_MEMPOOL_SIZE = Gauge(
    f"{METRIC_PREFIX}_mempool_size", "Number of unconfirmed transactions in mempool",
    labelnames=["blockchain"]
)
UTXO_NODE_MEMPOOL_USAGE = Gauge(f"{METRIC_PREFIX}_mempool_usage", "Total memory usage for the mempool", labelnames=["blockchain"])
UTXO_NODE_MEMPOOL_UNBROADCAST = Gauge(
    f"{METRIC_PREFIX}_mempool_unbroadcast", "Number of transactions waiting for acknowledgment", labelnames=["blockchain"]
)

UTXO_NODE_LATEST_BLOCK_HEIGHT = Gauge(
    f"{METRIC_PREFIX}_latest_block_height", "Height or index of latest block", labelnames=["blockchain"]
)
UTXO_NODE_LATEST_BLOCK_WEIGHT = Gauge(
    f"{METRIC_PREFIX}_latest_block_weight", "Weight of latest block according to BIP 141", labelnames=["blockchain"]
)
UTXO_NODE_LATEST_BLOCK_SIZE = Gauge(f"{METRIC_PREFIX}_latest_block_size", "Size of latest block in bytes", labelnames=["blockchain"])
UTXO_NODE_LATEST_BLOCK_TXS = Gauge(
    f"{METRIC_PREFIX}_latest_block_txs", "Number of transactions in latest block", labelnames=["blockchain"]
)

UTXO_NODE_TXCOUNT = Gauge(f"{METRIC_PREFIX}_txcount", "Number of TX since the genesis block", labelnames=["blockchain"])

UTXO_NODE_NUM_CHAINTIPS = Gauge(f"{METRIC_PREFIX}_num_chaintips", "Number of known blockchain branches", labelnames=["blockchain"])

UTXO_NODE_TOTAL_BYTES_RECV = Gauge(f"{METRIC_PREFIX}_total_bytes_recv", "Total bytes received", labelnames=["blockchain"])
UTXO_NODE_TOTAL_BYTES_SENT = Gauge(f"{METRIC_PREFIX}_total_bytes_sent", "Total bytes sent", labelnames=["blockchain"])

UTXO_NODE_LATEST_BLOCK_INPUTS = Gauge(
    f"{METRIC_PREFIX}_latest_block_inputs", "Number of inputs in transactions of latest block", labelnames=["blockchain"]
)
UTXO_NODE_LATEST_BLOCK_OUTPUTS = Gauge(
    f"{METRIC_PREFIX}_latest_block_outputs", "Number of outputs in transactions of latest block", labelnames=["blockchain"]
)
UTXO_NODE_LATEST_BLOCK_VALUE = Gauge(
    f"{METRIC_PREFIX}_latest_block_value", "Bitcoin value of all transactions in the latest block", labelnames=["blockchain"]
)
UTXO_NODE_LATEST_BLOCK_FEE = Gauge(
    f"{METRIC_PREFIX}_latest_block_fee", "Total fee to process the latest block", labelnames=["blockchain"]
)

UTXO_NODE_BAN_CREATED = Gauge(
    f"{METRIC_PREFIX}_ban_created", "Time the ban was created", labelnames=["address", "reason", "blockchain"]
)
UTXO_NODE_BANNED_UNTIL = Gauge(
    f"{METRIC_PREFIX}_banned_until", "Time the ban expires", labelnames=["address", "reason", "blockchain"]
)

UTXO_NODE_SERVER_VERSION = Gauge(f"{METRIC_PREFIX}_server_version", "The server version", labelnames=["blockchain"])
UTXO_NODE_PROTOCOL_VERSION = Gauge(f"{METRIC_PREFIX}_protocol_version", "The protocol version of the server", labelnames=["blockchain"])

UTXO_NODE_SIZE_ON_DISK = Gauge(f"{METRIC_PREFIX}_size_on_disk", "Estimated size of the block and undo files", labelnames=["blockchain"])

UTXO_NODE_VERIFICATION_PROGRESS = Gauge(
    f"{METRIC_PREFIX}_verification_progress", "Estimate of verification progress [0..1]", labelnames=["blockchain"]
)

UTXO_NODE_RPC_ACTIVE = Gauge(f"{METRIC_PREFIX}_rpc_active", "Number of RPC calls being processed", labelnames=["blockchain"])

EXPORTER_ERRORS = Counter(
    f"{METRIC_PREFIX}_exporter_errors", "Number of errors encountered by the exporter", labelnames=["type", "blockchain"]
)
PROCESS_TIME = Counter(
    f"{METRIC_PREFIX}_exporter_process_time", "Time spent processing metrics from node", labelnames=["blockchain"]
)

SATS_PER_COIN = Decimal(1e8)

def hashps_gauge_suffix(nblocks):
    if nblocks < 0:
        return "_neg%d" % -nblocks
    if nblocks == 120:
        return ""
    return "_%d" % nblocks


def hashps_gauge(num_blocks: int) -> Gauge:
    gauge = UTXO_NODE_HASHPS_GAUGES.get(num_blocks)
    if gauge is None:
        desc_end = "for the last %d blocks" % num_blocks
        if num_blocks == -1:
            desc_end = "since the last difficulty change"
        gauge = Gauge(
            f"{METRIC_PREFIX}_hashps%s" % hashps_gauge_suffix(num_blocks),
            "Estimated network hash rate per second %s" % desc_end,
            labelnames=["blockchain"]
        )
        UTXO_NODE_HASHPS_GAUGES[num_blocks] = gauge
    return gauge

def smartfee_gauge(num_blocks: int) -> Gauge:
    gauge = UTXO_NODE_ESTIMATED_SMART_FEE_GAUGES.get(num_blocks)
    if gauge is None:
        gauge = Gauge(
            f"{METRIC_PREFIX}_est_smart_fee_%d" % num_blocks,
            "Estimated smart fee per kilobyte for confirmation in %d blocks" % num_blocks,
            labelnames=["blockchain"]
        )
        UTXO_NODE_ESTIMATED_SMART_FEE_GAUGES[num_blocks] = gauge
    return gauge


config.py
import os
from distutils.util import strtobool

# Set up logging to look similar to bitcoin logs (UTC).
LOG_FORMAT = "%(asctime)s %(levelname)s %(message)s"

UTXO_NODE_BLOCKCHAIN_NAME = os.environ.get("UTXO_NODE_BLOCKCHAIN_NAME", "bitcoin")

UTXO_NODE_RPC_SCHEME = os.environ.get("UTXO_NODE_RPC_SCHEME", "http")
UTXO_NODE_RPC_HOST = os.environ.get("UTXO_NODE_RPC_HOST", "localhost")
UTXO_NODE_RPC_PORT = os.environ.get("UTXO_NODE_RPC_PORT", "8332")
UTXO_NODE_RPC_USER = os.environ.get("UTXO_NODE_RPC_USER")
UTXO_NODE_RPC_PASSWORD = os.environ.get("UTXO_NODE_RPC_PASSWORD")
UTXO_NODE_CONF_PATH = os.environ.get("UTXO_NODE_CONF_PATH")
HASHPS_BLOCKS = [int(b) for b in os.environ.get("HASHPS_BLOCKS", "-1,1,120").split(",") if b != ""]
SMART_FEES = [int(f) for f in os.environ.get("SMARTFEE_BLOCKS", "2,3,5,20").split(",") if f != ""]
METRICS_ADDR = os.environ.get("METRICS_ADDR", "")  # empty = any address
METRICS_PORT = int(os.environ.get("METRICS_PORT", "9332"))
RETRIES = int(os.environ.get("RETRIES", 5))
TIMEOUT = int(os.environ.get("TIMEOUT", 30))
RATE_LIMIT_SECONDS = int(os.environ.get("RATE_LIMIT", 5))
LOG_LEVEL = os.environ.get("LOG_LEVEL", "INFO")

FETCH_UPTIME = strtobool(os.environ.get("FETCH_UPTIME", 'True'))
FETCH_MEMINFO = strtobool(os.environ.get("FETCH_MEMINFO", 'True'))
FETCH_BLOCKCHAININFO = strtobool(os.environ.get("FETCH_BLOCKCHAININFO", 'True'))
FETCH_BLOCKSTATS = strtobool(os.environ.get("FETCH_BLOCKSTATS", 'True'))
FETCH_NETWORKINFO = strtobool(os.environ.get("FETCH_NETWORKINFO", 'True'))
FETCH_CHAINTIPS = strtobool(os.environ.get("FETCH_CHAINTIPS", 'True'))
FETCH_MEMPOOLINFO = strtobool(os.environ.get("FETCH_MEMPOOLINFO", 'True'))
FETCH_NETTOTALS = strtobool(os.environ.get("FETCH_NETTOTALS", 'True'))
FETCH_RPCINFO = strtobool(os.environ.get("FETCH_RPCINFO", 'True'))
FETCH_TXSTATS = strtobool(os.environ.get("FETCH_TXSTATS", 'True'))
FETCH_BANNED = strtobool(os.environ.get("FETCH_BANNED", 'True'))
FETCH_SMART_FEES = strtobool(os.environ.get("FETCH_SMART_FEES", 'True'))
FETCH_HASHP_BLOCKS = strtobool(os.environ.get("FETCH_HASHP_BLOCKS", 'True'))